finagle-core/src/main/scala/com/twitter/finagle/tracing/Id.scala

```diff
@@ -47,12 +47,33 @@ object SpanId {
 
   def apply(spanId: Long): SpanId = new SpanId(spanId)
 
-  def fromString(spanId: String): Option[SpanId] =
-    try {
-      Some(SpanId(new RichU64String(spanId).toU64Long))
-    } catch {
-      case NonFatal(_) => None
+  /**
+   * Parses a 1 to 32 character lower-hex string with no prefix into
+   * a 64-bit span id, by ignoring any bits higher than 64.
+   */
+  def fromString(spanId: String): Option[SpanId] = {
+    val length = spanId.length()
+    if (length < 1 || length > 32) {
+      return None
     }
+
+    // trim off any high bits
+    val startIndex = if (length > 16) length - 16 else 0
+
+    var result = 0L;
+    for (i <- startIndex until length) {
+      val c = spanId.charAt(i)
+      result <<= 4
+      if (c >= '0' && c <= '9') {
```

Comment:
again, performance related.. the code I pasted benchmarks better. you'll find that isDigit unfolds into quite many things. no problem reverting to RichU64String, if you say go!